design/ceph/object/object-store-backend.md

```diff
@@ -0,0 +1,103 @@
+---
+title: Object Store Backend
+target-version: release-1.10
+---
+
+# Overview
+
+Ceph RADOS Gateway is undergoing a major overhaul with the Zipper project. The goal of Project
+Zipper is to create an API that can be used to separate the RGW top end (represented by Swift and
+S3) from bottom end providers (represented currently by RADOS, and eventually by Cloud, File, and
+Memory backends). This should allow the front ends to access any implemented back end
+transparently.
+
+The secondary objective is to allow multiple layers of these API providers to stack up, allowing
+transformations of operations (such as redirection to different backends), or caching layers that
+cut across multiple backends.
+
+The primary Store is the current RadosStore, which provides access to a Ceph cluster via RADOS.
+However, new Stores are possible that store the data in any desired platform. One such Store, called
+DBStore, has been developed that stores data in SQL, and specifically in a local SQLite database.
+
+## Goals
+
+Rook-Ceph has a tremendous support for the Ceph Object Store and we want to leverage that.
+Especially the various controllers that are reconciling ObjectStore, Multisite configurations and
+Bucket provisioning.
+Mainly we are looking at providing the next generation of RGW via Rook.
+
+# Proposal details
+
+The `CephObjectStore` API will have a few details added to its spec such as an image so that it can
+be deployed without a Ceph cluster.
+
+```go
+type ObjectStoreSpec struct {
+...
+	// Image is the name of the image to use to deploy the object store
+	Image string `json:"image"`
+...
+}
+```
+
+Also, we need to expose a new type for the backend to use:
+
+```go
+type GatewaySpec struct {
+...
+	// Backend refers to the backend to use for the object store
+	// +optional
+	Backend *ObjectStoreBackend `json:"backend,omitempty"`
+...
+}
+```
+
+Where the `ObjectStoreBackend` is represented as such:
+
+```go
+type ObjectStoreBackend struct {
+	// The name of the backend
+	// +optional
+	// +kubebuilder:validation:Enum={"rados","sqlite"}
+	Name string `json:"name,omitempty"`
+
+    // Config is the configuration for the backend
+	// +optional
+    Config map[string]string `json:"config,omitempty"`
+}
+```
+
+Then the `CephObjectStore` `spec.go` file will need to be adapted depending on the backend.
+
+## Risks and Mitigation
+
+We still need to determine what the best approach between using the Rook-Ceph operator as is or
+building a new operator.
+
+Some pros to use the Rook-Ceph operator:
+* It is easier to use
+* Not much code changes involved
+* No need to change the various controller packages
+* When COSI is available we can integrate and deploy it
+* Code is maintained at a single location
+
+Some cons to use the Rook-Ceph operator:
```

Comment:
As I mentioned earlier if the backend is vendor-specific than a generic one. It may be difficult to implement in Rook

Reply:
That's a good point. We need to see what we can do in terms of extensibility here.